File: examples/cpp/velocity_publisher/src/velocity_publisher.cpp
#include "velocity_publisher.hpp"

#include <atomic>
#include <chrono>
#include <cmath>
#include <csignal>
#include <iomanip>
#include <iostream>
#include <thread>

constexpr auto FRAME_RATE = 10;
std::atomic_bool running{true};

void signalHandler(int) {
    std::cerr << "SIGINT or SIGTERM received. Exiting..." << std::endl;
    running = false;
}

int main(int argc, char* argv[]) {
    signal(SIGINT, signalHandler);
    signal(SIGTERM, signalHandler);

    nodar::zmq::VelocityPublisher publisher;

    std::cout << "Publishing velocity data at " << FRAME_RATE << " Hz" << std::endl;
    std::cout << "Press Ctrl+C to stop..." << std::endl;

    while (running) {
        // Get current timestamp
        const auto timestamp =
            std::chrono::duration_cast<std::chrono::nanoseconds>(std::chrono::system_clock::now().time_since_epoch())
                .count();

        // Example velocity data in odometry coordinate system
        // Odometry system: x=forward, y=left, z=up
        std::array<float, 3> velocity = {5.0f,  // vx: 5 m/s forward
                                         0.0f,  // vy: no lateral motion
                                         0.0f};  // vz: no vertical motion

        // Rotation matrix from odometry to Nodar raw camera coordinate system
        // Nodar raw camera system: x=right, y=down, z=forward
        // Transform: Nodar_x = -Odom_y, Nodar_y = -Odom_z, Nodar_z = Odom_x
        std::array<float, 9> rotationToNodarRaw = {0.0f, -1.0f, 0.0f,  // Row 1: Nodar x = -Odom y
                                                   0.0f, 0.0f,  -1.0f,  // Row 2: Nodar y = -Odom z
                                                   1.0f, 0.0f,  0.0f};  // Row 3: Nodar z = Odom x

        if (publisher.publishVelocity(timestamp, velocity, rotationToNodarRaw)) {
            std::cout << "\rPublishing | vx: " << std::fixed << std::setprecision(2) << velocity[0]
                      << " m/s, vy: " << velocity[1] << " m/s, vz: " << velocity[2] << " m/s" << std::flush;
        }

        std::this_thread::sleep_for(std::chrono::milliseconds((1000 / FRAME_RATE)));
    }

    std::cout << "\nPublisher stopped." << std::endl;
    return EXIT_SUCCESS;
}

File: examples/python/velocity_publisher/velocity_publisher/velocity_publisher.py
import os
import signal
import sys
import time
from datetime import datetime

import zmq

FRAME_RATE = 10

try:
    from zmq_msgs.velocity import VelocityData
    from zmq_msgs.topic_ports import VELOCITY_TOPIC
except ImportError:
    sys.path.append(
        os.path.abspath(os.path.join(os.path.dirname(__file__), "../../../../zmq_msgs/python"))
    )
    from zmq_msgs.velocity import VelocityData
    from zmq_msgs.topic_ports import VELOCITY_TOPIC

# Global variable for signal handling
running = True


def signal_handler(signum, frame):
    global running
    print("\nSIGINT or SIGTERM received. Exiting...", file=sys.stderr)
    running = False


class Publisher:
    def __init__(self, topic, ip, port):
        self.context = zmq.Context(1)
        self.socket = self.context.socket(zmq.PUB)
        self.socket.setsockopt(zmq.SNDHWM, 1)
        if ip:
            self.socket.connect(f"tcp://{ip}:{port}")
        else:
            self.socket.bind(f"tcp://*:{port}")

    def send(self, buffer):
        self.socket.send(buffer)


class VelocityPublisher:
    def __init__(self):
        self.publisher = Publisher(VELOCITY_TOPIC.name, "", VELOCITY_TOPIC.port)
        time.sleep(1)

    def publish_velocity(self, timestamp, velocity, rotation_to_nodar_raw):
        """
        Publish velocity data.

        Args:
            timestamp: timestamp in nanoseconds
            velocity: list/tuple of 3 floats [vx, vy, vz] in m/s
            rotation_to_nodar_raw: list/tuple of 9 floats representing 3x3 rotation matrix
        """
        velocity_data = VelocityData(timestamp, velocity, rotation_to_nodar_raw)
        buffer = bytearray(velocity_data.msg_size())
        velocity_data.write(buffer, 0)
        self.publisher.send(buffer)
        return True


def main():
    global running

    # Setup signal handlers
    signal.signal(signal.SIGINT, signal_handler)
    signal.signal(signal.SIGTERM, signal_handler)

    publisher = VelocityPublisher()

    print(f"Publishing velocity data at {FRAME_RATE} Hz")
    print("Press Ctrl+C to stop...")

    while running:
        # Get current timestamp
        timestamp = int(datetime.now().timestamp() * 1e9)

        # Example velocity data in odometry coordinate system
        # Odometry system: x=forward, y=left, z=up
        velocity = [
            5.0,  # vx: 5 m/s forward
            0.0,  # vy: no lateral motion
            0.0   # vz: no vertical motion
        ]

        # Rotation matrix from odometry to Nodar raw camera coordinate system
        # Nodar raw camera system: x=right, y=down, z=forward
        # Transform: Nodar_x = -Odom_y, Nodar_y = -Odom_z, Nodar_z = Odom_x
        rotation_to_nodar_raw = [
            0.0, -1.0, 0.0,  # Row 1: Nodar x = -Odom y
            0.0, 0.0, -1.0,  # Row 2: Nodar y = -Odom z
            1.0, 0.0, 0.0    # Row 3: Nodar z = Odom x
        ]

        if publisher.publish_velocity(timestamp, velocity, rotation_to_nodar_raw):
            print(
                f"\rPublishing | vx: {velocity[0]:.2f} m/s, vy: {velocity[1]:.2f} m/s, vz: {velocity[2]:.2f} m/s",
                end="",
                flush=True
            )

        time.sleep(1.0 / FRAME_RATE)

    print("\nPublisher stopped.")


if __name__ == "__main__":
    main()